windows/win_chocolatey.ps1

```diff
@@ -96,153 +83,261 @@ Else
+    )
+
+    $cmd = "$executable list --local-only $package"
+    $results = invoke-expression $cmd
+
+    if ($LastExitCode -ne 0)
+    {
+        Set-Attr $result "choco_error_cmd" $cmd
+        Set-Attr $result "choco_error_log" "$results"
+        
+        Throw "Error checking installation status for $package" 
+    } 
+    
+    If ("$results" -match " $package .* (\d+) packages installed.")
+    {
+        return $matches[1] -gt 0
+    }
+    
+    $false
 }
 
-if ($source -eq "webpi")
+Function Choco-Upgrade 
 {
-    # check whether 'webpi' installation source is available; if it isn't, install it
-    $webpi_check_cmd = "$executable list webpicmd -localonly"
-    $webpi_check_result = invoke-expression $webpi_check_cmd
-    Set-Attr $result "chocolatey_bootstrap_webpi_check_cmd" $webpi_check_cmd
-    Set-Attr $result "chocolatey_bootstrap_webpi_check_log" $webpi_check_result
-    if (
-        (
-            ($webpi_check_result.GetType().Name -eq "String") -and
-            ($webpi_check_result -match "No packages found")
-        ) -or
-        ($webpi_check_result -contains "No packages found.")
+    [CmdletBinding()]
+    
+    param(
+        [Parameter(Mandatory=$true, Position=1)]
+        [string]$package,
+        [Parameter(Mandatory=$false, Position=2)]
+        [string]$version,
+        [Parameter(Mandatory=$false, Position=3)]
+        [string]$source,
+        [Parameter(Mandatory=$false, Position=4)]
+        [bool]$force
     )
+
+    if (-not (Choco-IsInstalled $package))
     {
-        #lessmsi is a webpicmd dependency, but dependency resolution fails unless it's installed separately
-        $lessmsi_install_cmd = "$executable install lessmsi"
-        $lessmsi_install_result = invoke-expression $lessmsi_install_cmd
-        Set-Attr $result "chocolatey_bootstrap_lessmsi_install_cmd" $lessmsi_install_cmd
-        Set-Attr $result "chocolatey_bootstrap_lessmsi_install_log" $lessmsi_install_result
+        throw "$package is not installed, you cannot upgrade"
+    }
 
-        $webpi_install_cmd = "$executable install webpicmd"
-        $webpi_install_result = invoke-expression $webpi_install_cmd
-        Set-Attr $result "chocolatey_bootstrap_webpi_install_cmd" $webpi_install_cmd
-        Set-Attr $result "chocolatey_bootstrap_webpi_install_log" $webpi_install_result
+    $cmd = "$executable upgrade -y $package"
 
-        if (($webpi_install_result | select-string "already installed").length -gt 0)
-        {
-            #no change
-        }
-        elseif (($webpi_install_result | select-string "webpicmd has finished successfully").length -gt 0)
+    if (-not $source)
+    {
+        $cmd += " -source https://chocolatey.org/api/v2/"
+    }
+    elseif ($source -eq "chocolatey")
+    {
+        $cmd += " -source https://chocolatey.org/api/v2/"
+    }
+    elseif (($source -eq "windowsfeatures") -or `
+        ($source -eq "webpi") -or ($source -eq "ruby"))
+    {
+        $cmd += " -source $source"
```

Comment:
None of the alternative feeds work in 0.9.9. They will be back in 0.9.10 - see https://github.com/chocolatey/choco/issues/14